### PR TITLE
fix: Multiple imports from custom packages generated in separate lines

### PR DESCRIPTION
**Pull request by Savio629**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: https://github.com/SamagraX-Stencil/stencil-cli/issues/11

## Does this PR introduce a breaking change?
- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by Savio629** [comment]
// @sooraj1002 @techsavvyash

**Comment by Savio629** [comment]
@sooraj1002

File: src/utils/module-import.declarator.ts [hunk 1]
```diff
@@ -9,10 +9,35 @@ export class ModuleImportDeclarator {
     const toInsert = this.buildLineToInsert(options);
     const contentLines = content.split('\n');
     const finalImportIndex = this.findImportsEndpoint(contentLines);
+    const updatedContent = this.mergeImportStatements(contentLines, finalImportIndex, toInsert,options);
+    return updatedContent.join('\n');
+}
+
+private mergeImportStatements(contentLines: string[], finalImportIndex: number, toInsert: string, options: DeclarationOptions): string[] {
+  const existingImportLineIndex = contentLines.findIndex(line =>
+    line.startsWith(`import {`) && line.includes(`} from '${options.path}';`)
```
Comment (techsavvyash) [inline src/utils/module-import.declarator.ts]: can we turn this to a regex?

**Review by techsavvyash — CHANGES_REQUESTED** [review]

File: src/utils/module-import.declarator.ts (hunk 1 above)
Comment (techsavvyash) [reply src/utils/module-import.declarator.ts]: also, could you please share a small video showing that this works?

Comment (techsavvyash) [reply src/utils/module-import.declarator.ts]: also, please include tests in the cli to test out this functionality.

**Comment by Savio629** [comment]
@techsavvyash 

https://github.com/SamagraX-Stencil/schematics/assets/91362589/6150811d-fd03-4a22-a9f1-43c32c476377

**Comment by Savio629** [comment]
@techsavvyash Can you specify a bit more on what type of test should I write?

**Comment by techsavvyash** [comment]
> @techsavvyash Can you specify a bit more on what type of test should I write?

try to think of a few cases which this PR must be solving.
for example:
1. What if a single import from the `@samagra-x/stencil` pacakge is already present in the file and you need to add a new import from the same package, will it club the imports in that case? 
2. The demo in the video can be a test case in itself. 
3. Test with different quote chars like `'` and `"` in case of already present imports like in case 1.
4. Does the PR support adding of imports with alias such as `import { foo as bar } from 'foobar'`

**Comment by Savio629** [comment]
@techsavvyash When I run npm run test there are some test which fail already before my code changes
image

**Comment by techsavvyash** [comment]
Could you update the tests?

**Comment by Savio629** [comment]
I'm not good with test but I'll try
can you specify me how can i run a specific test file
`npx jest test/utils/module-import.declarator.test.ts`
is this the right way?
UPDATE: running this command solves the problem `npx jest test/utils/module-import.declarator.test.ts --config jest.json`
image

**Comment by Savio629** [comment]
@techsavvyash 

https://github.com/SamagraX-Stencil/schematics/assets/91362589/e4b7789c-2599-4051-9c0c-e997cc4c6478

The code works well for clubbing imports and also for clubbing imports even those alias is present but fails for 
Test with different quote chars like ' and " in case of already present imports like in case 1.

Am I doing something wrong here or I'll have to fix for quotes case from the code itself?
This are the test that I wrote

``` 
it('should not break if same package already exists', () => {
  const content: string =
      "import { Module } from '@nestjs/common';\n" +
      "import { TestModule } from '@samagra-x/stencil';\n" +
      '\n' +
      '@Module({})\n' +
      'const x = " from ";\n' +
      'console.error(" from ");\n' +
      'export class FooModule {}\n';
    const options: DeclarationOptions = {
      metadata: 'imports',
      name: 'module',
      path: normalize('@samagra-x/stencil'),
      module: normalize('/src/foo/foo.module.ts'),
      symbol: 'Foo',
      isPackage: false
    };
    const declarator = new ModuleImportDeclarator();
    expect(declarator.declare(content, options)).toEqual(
      "import { Module } from '@nestjs/common';\n" +
      "import { TestModule, Foo} from '@samagra-x/stencil';\n" +
        '\n' +
        '@Module({})\n' +
        'const x = " from ";\n' +
        'console.error(" from ");\n' +
        'export class FooModule {}\n',
    );
  });

  it('should not break for different quote chars', () => {
    const content: string =
        "import { Module } from '@nestjs/common';\n" +
        "import { TestModule } from \"@samagra-x/stencil\" ;\n" +
        '\n' +
        '@Module({})\n' +
        'const x = " from ";\n' +
        'console.error(" from ");\n' +
        'export class FooModule {}\n';
      const options: DeclarationOptions = {
        metadata: 'imports',
        name: '@samagra-x/stencil',
        path: normalize('src/foo'),
        module: normalize('/src/foo/foo.module.ts'),
        symbol: 'Foo',
        isPackage: false
      };
      const declarator = new ModuleImportDeclarator();
      expect(declarator.declare(content, options)).toEqual(
        "import { Module } from '@nestjs/common';\n" +
        "import { TestModule, Foo } from \"@samagra-x/stencil\";\n" +
          '\n' +
          '@Module({})\n' +
          'const x = " from ";\n' +
          'console.error(" from ");\n' +
          'export class FooModule {}\n',
      );
    });

    it('should not break if alias is present', () => {
      const content: string =
          "import { Module } from '@nestjs/common';\n" +
          "import { TestModule } from '@samagra-x/stencil';\n" +
          '\n' +
          '@Module({})\n' +
          'const x = " from ";\n' +
          'console.error(" from ");\n' +
          'export class FooModule {}\n';
        const options: DeclarationOptions = {
          metadata: 'imports',
          name: 'module',
          path: normalize('@samagra-x/stencil'),
          module: normalize('/src/foo/foo.module.ts'),
          symbol: 'Foo as F',
          isPackage: false
        };
        const declarator = new ModuleImportDeclarator();
        expect(declarator.declare(content, options)).toEqual(
          "import { Module } from '@nestjs/common';\n" +
          "import { TestModule, Foo as F} from '@samagra-x/stencil';\n" +
            '\n' +
            '@Module({})\n' +
            'const x = " from ";\n' +
            'console.error(" from ");\n' +
            'export class FooModule {}\n',
        );
      });  
```

**Comment by techsavvyash** [comment]
Screenshot 2024-04-18 at 12 01 46 PM

why is a `./` coming in front of the package name? I think this might be the reason that the imports not being clubbed together, though it maybe a code level issue in your implementation as well, but not very likely because the schematics in itself are stable enough.

**Comment by Savio629** [comment]
I too am confused
Can you check if the below test is right?
```
 it('should not break for different quote chars', () => {
    const content: string =
        "import { Module } from '@nestjs/common';\n" +
        "import { TestModule } from \"@samagra-x/stencil\" ;\n" +
        '\n' +
        '@Module({})\n' +
        'const x = " from ";\n' +
        'console.error(" from ");\n' +
        'export class FooModule {}\n';
      const options: DeclarationOptions = {
        metadata: 'imports',
        name: '@samagra-x/stencil',
        path: normalize('src/foo'),
        module: normalize('/src/foo/foo.module.ts'),
        symbol: 'Foo',
        isPackage: false
      };
      const declarator = new ModuleImportDeclarator();
      expect(declarator.declare(content, options)).toEqual(
        "import { Module } from '@nestjs/common';\n" +
        "import { TestModule, Foo } from \"@samagra-x/stencil\";\n" +
          '\n' +
          '@Module({})\n' +
          'const x = " from ";\n' +
          'console.error(" from ");\n' +
          'export class FooModule {}\n',
      );
    });
```

**Comment by Savio629** [comment]
@techsavvyash  I have now fix for different quote chars
I have pushed the test file too
You can test the test file using the below command
`npx jest module-import.declarator.test.ts --config jest.json`

**Review by techsavvyash — APPROVED** [review]
LGTM

**Pull request merged**